### DISCUSSION TITLE
Git repositories using SHA256 object format break helix_vcs features

### DISCUSSION THREAD

**Discussion by bvergnaud** [helix-editor/helix#12117] [General]
### Summary

In a git repository that uses SHA256 object format, the gutter in helix does not indicate additions/changes/deletions as it normally does, and resetting diff patches does not work with helix indicating that Diff is not available in the buffer.

It seems that [gix](https://github.com/Byron/gitoxide) does not support SHA256 yet which is the most likely explanation of the problem. It causes a similar issue in [starship](https://github.com/starship/starship) where the prompt does not show the git related informations. It also means that there might be other problems in such repositories that I haven't yet noticed. I don't have an exhaustive view of all the ways helix relies on git, but at least ignored files are still ignored properly.

Possible solutions would be:
* Wait for gix to implement support for SHA256 object format. [There's an issue for it, but no timeline](https://github.com/Byron/gitoxide/issues/281).
* Switch to a different implementation. git2 is [not there yet either](https://github.com/rust-lang/git2-rs/issues/1090), I don't know if there are other solid git crates.
* Use custom code / build a new git library. Seems overkill but it's possible nonetheless. :D 

### Reproduction Steps

A simple way to replicate the issue:
```sh
for obj_format in sha1 sha256 ; do
  git init --object-format=sha1 "test-${obj_format}"
  printf "this\nis\na\ntest\n" > "test-${obj_format}/file.txt"
  git -C "test-${obj_format}/" add file.txt
  git -C "test-${obj_format}/" commit -m "initial commit"
done
```

Then compare editing `file.txt` in each test repository by adding / changing / removing lines and see that the gutter shows nothing in the test-sha256 repository.

### Helix log

```
2024-11-23T19:29:42.333 helix_vcs [DEBUG] Error {
    context: "failed to open git repo",
    source: Open(
        Config(
            ConfigTypedString(
                Error {
                    key: "extensions.objectFormat",
                    value: Some(
                        "sha256",
                    ),
                    environment_override: None,
                    source: None,
                },
            ),
        ),
    ),
}
2024-11-23T19:29:42.333 helix_vcs [DEBUG] failed to open diff base for /home/bvergnaud/Downloads/helix-git-gutter-bug-demo/test-sha256/file.txt
2024-11-23T19:29:42.333 helix_vcs [DEBUG] Error {
    context: "failed to open git repo",
    source: Open(
        Config(
            ConfigTypedString(
                Error {
                    key: "extensions.objectFormat",
                    value: Some(
                        "sha256",
                    ),
                    environment_override: None,
                    source: None,
                },
            ),
        ),
    ),
}
2024-11-23T19:29:42.333 helix_vcs [DEBUG] failed to obtain current head name for /home/bvergnaud/Downloads/helix-git-gutter-bug-demo/test-sha256/file.txt
```

### Platform

Linux

### Terminal Emulator

kitty

### Installation Method

Official extra/helix package in arch linux repositories

### Helix Version

helix 24.7

**Comment by pascalkuthe** [comment]
Sh256 support is not widespread as you have noticed. If that's important to you, you could take a look at contributing that to gitoxide. Otherwise a more pragmatic solution is to use sha256. We will not switch git implementation (even if git2 did support it).

**Reply by bvergnaud** [reply]
That's fair and expected, I opened the issue mostly for tracking, it's nowhere near critical for now.
I'll keep an eye on the gitoxide issue. Thanks for the quick answer. :)
